### DISCUSSION TITLE
Add support for the "using" keyword and resource management

### DISCUSSION THREAD

**Discussion by jdgriffith** [oven-sh/bun#4325] [Ideas]
For reference: https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#using-declarations-and-explicit-resource-management

> That brings us to the first star of the feature: using declarations! using is a new keyword that lets us declare new fixed bindings, kind of like const. The key difference is that variables declared with using get their Symbol.dispose method called at the end of the scope

Examples:

```
export function doSomeWork() {
    using file = new TempFile(".some_temp_file");

    // use file...

    if (someCondition()) {
        // do some more work...
        return;
    }
}
```

And `await using ...`

```
async function func() {
    await using a = loggy("a");
    await using b = loggy("b");
    {
        await using c = loggy("c");
        await using d = loggy("d");
    }
    await using e = loggy("e");
    return;

    // Unreachable.
    // Never created, never disposed.
    await using f = loggy("f");
}
```

If this is already supported, then awesome! But if not, I feel this will be a key feature going forward for those of us concerned about performance and having control (in the least amount of code) of memory and such.

**Comment by okikio** [comment]
Is there any progress on this?

**Reply by Jarred-Sumner** [reply]
yes, it was implemented awhile ago

**Reply by okikio** [reply]
Oh wow now that's fast

**Comment by Jarred-Sumner** [comment]
Implemented in [Bun v1.0.23 ](https://bun.sh/blog/bun-v1.0.23#resource-management-is-now-supported)
